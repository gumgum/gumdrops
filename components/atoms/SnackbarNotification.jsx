import React, { Component } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

class SnackbarNotification extends Component {
    static propTypes = {
        /** Unique key (string or number) for identification, *Required */
        id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
        /** Message or content of the notification */
        text: PropTypes.string,
        /** Context of notification (success, info, warning, danger, hidden) */
        context: PropTypes.string,
        /** Time in ms to hide notification, if 0 is passed ignores timeout */
        msToClose: PropTypes.number,
        /** Hides the close button, if true auto enables timeout to hide notification */
        hideCloseButton: PropTypes.bool,
        /** Function to be called after hiding notification */
        callback: PropTypes.func,
        /** Component children that can be used as content */
        children: PropTypes.node,
        /** Styles passed through props */
        style: PropTypes.object,
        /** Class names passed through props */
        className: PropTypes.string
    };

    static defaultProps = {
        msToClose: 3000,
        hideCloseButton: false
    };

    displayName = 'SnackbarNotification';

    state = {
        shouldHide: false // flag to know if component should be visible
    };

    componentDidMount() {
        // Set the timeout to hide component if applies
        this._setTimeout();
    }

    componentWillUnmount() {
        // Clears timeout if exists
        this._clearTimeout();
    }

    _timeout = null;

    _setTimeout = () => {
        const { msToClose, hideCloseButton } = this.props;
        // If close button is not showing always set timeout
        // If msToClose is 0 skip timeout
        const timeout =
            hideCloseButton || msToClose !== 0 ? setTimeout(this._closeSnackbar, msToClose) : null;
        // Sets timeout in property if it wasn't setted before
        if (!this._timeout) this._timeout = timeout;
    };

    _clearTimeout = () => {
        if (this._timeout) clearTimeout(this._timeout);
    };

    // Sets the hide in state and calls callback function
    _closeSnackbar = () => {
        // Clear timeout if still running
        this._clearTimeout();
        // If there's a callback function pass the id
        const { callback, id } = this.props;
        const cb = callback ? () => callback(id) : null;
        this.setState(
            {
                shouldHide: true
            },
            cb
        );
    };

    render() {
        const { text, context, hideCloseButton, className, style, children } = this.props;
        // Gets the content of the snackbar, either children or message passed through props
        // If there's no content, don't render the component
        const content = children || text;
        if (!content) {
            return null;
        }

        // Sets the class of the component depending on the prop.context value
        const baseClass = 'gds-snackbar__notification';
        const rootClass = cx(baseClass, className, {
            [`${baseClass}--hidden`]: this.state.shouldHide,
            [`${baseClass}--${context}`]: context
        });

        const buttonClass = cx(`${baseClass}-button`, {
            [`${baseClass}-button--${context}`]: context
        });

        return (
            <li className={rootClass} style={style}>
                <p className="gds-snackbar__notification-text">{content}</p>
                {hideCloseButton || (
                    <button className={buttonClass} onClick={this._closeSnackbar} />
                )}
            </li>
        );
    }
}

export default SnackbarNotification;
