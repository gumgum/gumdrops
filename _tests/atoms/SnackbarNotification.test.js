/* globals shallow mount */
import React from 'react';
import SnackbarNotification from '../../components/atoms/SnackbarNotification';

const noop = () => {};

const props = {
    id: 1,
    text: 'Snackbar Message',
    context: 'info',
    callback: noop
};

describe('Expect <SnackbarNotification>', () => {
    it('to render the component', () => {
        const wrapper = shallow(<SnackbarNotification {...props} />);
        expect(wrapper.find('.gds-snackbar__notification').length).toBe(1);
        expect(wrapper.text()).toMatch('Snackbar Message');
        expect(wrapper).toMatchSnapshot();
    });

    it('to render a child component', () => {
        const wrapper = shallow(
            <SnackbarNotification id="1">Snackbar Message</SnackbarNotification>
        );
        expect(wrapper.find('p.gds-snackbar__notification-text').length).toBe(1);
        expect(wrapper.find('p.gds-snackbar__notification-text').text()).toMatch(
            'Snackbar Message'
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('to apply the given context as a css class', () => {
        const wrapper = shallow(<SnackbarNotification {...props} context="warning" />);
        expect(wrapper.find('.gds-snackbar__notification--warning').length).toBe(1);
        expect(wrapper.find('.gds-snackbar__notification-button--warning').length).toBe(1);
        expect(wrapper).toMatchSnapshot();
    });

    it('to not render if there is no message', () => {
        const wrapper = shallow(<SnackbarNotification id="1" />);
        expect(wrapper.find('.gds-snackbar__notification').length).toBe(0);
        expect(wrapper).toMatchSnapshot();
    });

    it('to be hidden when close button is pressed', () => {
        const wrapper = shallow(<SnackbarNotification {...props} />);
        const closeBtn = wrapper.find('.gds-snackbar__notification-button');
        expect(wrapper.find('.gds-snackbar__notification--hidden').length).toBe(0);
        closeBtn.simulate('click');
        expect(wrapper.find('.gds-snackbar__notification--hidden').length).toBe(1);
        expect(wrapper).toMatchSnapshot();
    });

    it('to call callback after hide', () => {
        const mockCallback = jest.fn();
        const wrapper = shallow(<SnackbarNotification {...props} callback={mockCallback} />);
        const closeBtn = wrapper.find('.gds-snackbar__notification-button');
        closeBtn.simulate('click');
        expect(mockCallback).toBeCalledTimes(1);
        expect(wrapper).toMatchSnapshot();
    });

    it('to hide after 2 seconds', () => {
        jest.useFakeTimers();
        const callback = jest.fn();
        const wrapper = shallow(
            <SnackbarNotification {...props} msToClose={2000} callback={callback} />
        );
        jest.advanceTimersByTime(1000);
        expect(wrapper.find('.gds-snackbar__notification--hidden').length).toBe(0);
        expect(callback).toHaveBeenCalledTimes(0);
        jest.advanceTimersByTime(2500);
        // Calling update on wrapper to solve this issue:
        // https://github.com/airbnb/enzyme/issues/1229
        wrapper.update();
        expect(callback).toHaveBeenCalledTimes(1);
        expect(wrapper.find('.gds-snackbar__notification--hidden').length).toBe(1);
    });

    it('to not show close button', () => {
        const wrapper = shallow(<SnackbarNotification {...props} hideCloseButton />);
        expect(wrapper.find('.gds-snackbar__notification-button').length).toBe(0);
    });

    it('to not autoclose', () => {
        jest.useFakeTimers();
        const wrapper = shallow(<SnackbarNotification {...props} msToClose={0} />);
        expect(wrapper.find('.gds-snackbar__notification--hidden').length).toBe(0);
        jest.advanceTimersByTime(5000);
        wrapper.update();
        expect(wrapper.find('.gds-snackbar__notification--hidden').length).toBe(0);
    });
});
