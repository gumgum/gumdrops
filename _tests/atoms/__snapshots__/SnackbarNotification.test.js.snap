// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect <SnackbarNotification> to apply the given context as a css class 1`] = `
<li
  className="gds-snackbar__notification gds-snackbar__notification--warning"
>
  <p
    className="gds-snackbar__notification-text"
  >
    Snackbar Message
  </p>
  <button
    className="gds-snackbar__notification-button gds-snackbar__notification-button--warning"
    onClick={[Function]}
  />
</li>
`;

exports[`Expect <SnackbarNotification> to be hidden when close button is pressed 1`] = `
<li
  className="gds-snackbar__notification gds-snackbar__notification--hidden gds-snackbar__notification--info"
>
  <p
    className="gds-snackbar__notification-text"
  >
    Snackbar Message
  </p>
  <button
    className="gds-snackbar__notification-button gds-snackbar__notification-button--info"
    onClick={[Function]}
  />
</li>
`;

exports[`Expect <SnackbarNotification> to call callback after hide 1`] = `
<li
  className="gds-snackbar__notification gds-snackbar__notification--hidden gds-snackbar__notification--info"
>
  <p
    className="gds-snackbar__notification-text"
  >
    Snackbar Message
  </p>
  <button
    className="gds-snackbar__notification-button gds-snackbar__notification-button--info"
    onClick={[Function]}
  />
</li>
`;

exports[`Expect <SnackbarNotification> to not render if there is no message 1`] = `""`;

exports[`Expect <SnackbarNotification> to render a child component 1`] = `
<li
  className="gds-snackbar__notification"
>
  <p
    className="gds-snackbar__notification-text"
  >
    Snackbar Message
  </p>
  <button
    className="gds-snackbar__notification-button"
    onClick={[Function]}
  />
</li>
`;

exports[`Expect <SnackbarNotification> to render the component 1`] = `
<li
  className="gds-snackbar__notification gds-snackbar__notification--info"
>
  <p
    className="gds-snackbar__notification-text"
  >
    Snackbar Message
  </p>
  <button
    className="gds-snackbar__notification-button gds-snackbar__notification-button--info"
    onClick={[Function]}
  />
</li>
`;
